### PR TITLE
qtcreator: 5.0.3 -> 9.0.1

### LINKED ISSUE #173751: qtcreator outdated
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: **qtcreator**

current version: 5.0.3
desired version: 7.0.1

###### Notify maintainers

maintainers: @akaWolf 

###### Note for maintainers

Please tag this issue in your PR.

### LINKED ISSUE #197006: qtcreator is too old
###### Project name
`nix search` name:
qtcreator
current version: 5.0.3
desired version: 8.0.1

###### Notify maintainers

maintainers: @akaWolf 

###### Note for maintainers

qtcreator 8.0.1 needs qt6, but it add many features, included ui promote. So I think maybe it should be update?

### PR DESCRIPTION
**Pull request by milahu**
#### Description of changes

fix #173751
fix #197006

#### TODO

* fix segfault
  * enable debug build with `cmakeBuildType = "Debug";`
  * run `QT_LOGGING_RULES='*.debug=true' qtcreator` to see `qDebug` messages
  * fix `spec->d->kill();`
  * fix loading of `QmlProjectPlugin`
  * fix loading of plugins
  * upstream issue https://bugreports.qt.io/browse/QTCREATORBUG-28132
* test `clang`
  * do we need [qt-clang](https://code.qt.io/cgit/clang/llvm-project.git/)?
  * do we need [clazy](https://code.qt.io/cgit/clang/clazy.git/)?
    * https://github.com/KDE/clazy Qt-oriented static code analyzer based on the Clang framework
* use clang 14
* add qtwebengine. why does it rebuild qtwebengine on my machine?
* do we need [litehtml](https://github.com/litehtml/litehtml)? to generate docs?

https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/README.md

#### segfault

```
#0  0x00007fffdb4eb56c in QmlProjectManager::Internal::QmlProjectPlugin::~QmlProjectPlugin() ()
   from /nix/store/kgg4lvi7hfy2zaph1r940jvpl4pd83v4-qtcreator-8.0.1/lib/qtcreator/plugins/libQmlProjectManager.so
#1  0x00007fffdb4eb689 in QmlProjectManager::Internal::QmlProjectPlugin::~QmlProjectPlugin() ()
   from /nix/store/kgg4lvi7hfy2zaph1r940jvpl4pd83v4-qtcreator-8.0.1/lib/qtcreator/plugins/libQmlProjectManager.so
#2  0x00007ffff7f9b016 in ExtensionSystem::Internal::PluginSpecPrivate::kill() ()
   from /nix/store/kgg4lvi7hfy2zaph1r940jvpl4pd83v4-qtcreator-8.0.1/bin/../lib/qtcreator/libExtensionSystem.so.8
#3  0x00007ffff7f90237 in ExtensionSystem::Internal::PluginManagerPrivate::loadPlugins() ()
   from /nix/store/kgg4lvi7hfy2zaph1r940jvpl4pd83v4-qtcreator-8.0.1/bin/../lib/qtcreator/libExtensionSystem.so.8
#4  0x000000000040ca28 in main ()
```

seems not related to #177720
because dropping the cache has no effect:
`find ${XDG_CACHE_HOME:-$HOME/.cache}/**/qmlcache -type f -delete`

looks like a double-free in `~QmlProjectPlugin()`

```cc
// qt-creator-opensource-src-8.0.1/src/libs/extensionsystem/pluginmanager.cpp

void PluginManagerPrivate::loadPlugins()
{
    const QVector<PluginSpec *> queue = loadQueue();
    Utils::setMimeStartupPhase(MimeStartupPhase::PluginsLoading);
    for (PluginSpec *spec : queue)
        loadPlugin(spec, PluginSpec::Loaded);

    Utils::setMimeStartupPhase(MimeStartupPhase::PluginsInitializing);
    for (PluginSpec *spec : queue)
        loadPlugin(spec, PluginSpec::Initialized);

    Utils::setMimeStartupPhase(MimeStartupPhase::PluginsDelayedInitializing);
    Utils::reverseForeach(queue, [this](PluginSpec *spec) {
        loadPlugin(spec, PluginSpec::Running);
        if (spec->state() == PluginSpec::Running) {
            delayedInitializeQueue.push(spec);
        } else {
            // Plugin initialization failed, so cleanup after it
            spec->d->kill();
        }
    });
    emit q->pluginsChanged();
    // ...
}
```

#### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/qtcreator/default.nix [hunk 1]
```diff
@@ -1,84 +1,82 @@
...
+    pkg-config
+    wrapQtAppsHook
+  ];
 
   doCheck = true;
 
+  postPatch = ''
+    stat src/libs/extensionsystem/pluginmanager.cpp
+    cp ${./src/qt-creator-opensource-src-8.0.1/src/libs/extensionsystem/pluginmanager.cpp} src/libs/extensionsystem/pluginmanager.cpp
+    cp ${./src/qt-creator-opensource-src-8.0.1/src/plugins/welcome/welcomeplugin.cpp} src/plugins/welcome/welcomeplugin.cpp
```
Comment (SuperSandro2000) [inline pkgs/development/tools/qtcreator/default.nix:68]: Can we fetch those files from somewhere instead?

Comment (milahu) [reply pkgs/development/tools/qtcreator/default.nix:68]: this is for debugging, will remove later

**Comment by andrevmatos** [comment]
[qtcreator 9 is out today](https://www.qt.io/blog/qt-creator-9-released)

**Comment by puetzk** [comment]
> do we need [qt-clang](https://code.qt.io/cgit/clang/llvm-project.git/)?

Generally no, this is mostly a mirror. Though for any given release Qt will often have cherry-picked some specific  fixes they felt were worth it, or that they have prosed but weren't merged yet. But it tends to be more in the scope of plus or minus fixes than being a major fork.

e.g. https://code.qt.io/cgit/clang/llvm-project.git/log/?h=release_15.0.0-based  diverges after 
4ba6a9c9f65bbc8bd06e3652cb20fd4dfc846137, taking 3 more patches that missed upstream's [release/15.x](https://github.com/llvm/llvm-project/compare/release/15.x)
- https://code.qt.io/cgit/clang/llvm-project.git/commit/?h=release_15.0.0-based&id=a6ac41d93949610423bd04c2aae9f4e0dadbe0b9
  only really important on windows, since an opened file handle does not block deletion/overwrite on linux.
- https://code.qt.io/cgit/clang/llvm-project.git/commit/?h=release_15.0.0-based&id=5beadd4ba7929a8b94df8c99c621063de9bccd10
  workaround for building clang separately (against a system-installed rest of llvm). So (if I understand it correctly) only important if you are building a separate clang, but not a itself a reason to do so.
- https://code.qt.io/cgit/clang/llvm-project.git/commit/?h=release_15.0.0-based&id=218224bce00c69b6f8a396e681c51256fc6f1b99
  probably the most relevant, it's a fairly-stale upstream proposal https://reviews.llvm.org/D53072 that Qt Creator has been carrying around for some time. Changes the behavior of clang-format, but shouldn't make or break the whole IDE.

> do we need [clazy](https://code.qt.io/cgit/clang/clazy.git/)?

clazy is certainly a useful static analysis tool in Qt/KDE code. But it's should be more an additional nice-to-have than a blocker for updating qtcreator.

**Comment by puetzk** [comment]
> do we need [litehtml](https://github.com/litehtml/litehtml)? to generate docs?

Not to generate them, but rather to view them in the IDE's help window or context-sensntive help. By default QtCreator's build system should bring in https://code.qt.io/cgit/playground/qlitehtml.git/ as a git submodule and build it as part of building qt creator, but that should happily use a system litehtml (which nixpkgs appears to already have: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/litehtml/default.nix) as long as find_package works.

https://github.com/qt-creator/qt-creator/blob/master/.gitmodules#L12-L14

**Comment by milahu** [comment]
did some stabbing at qtcreator 9.0.1 in https://github.com/milahu/nixpkgs/tree/qtcreator-8.0.1-debug

```console
$ qtcreator
Warning: No themes found in installation:  "/share/qtcreator/themes"
Failed to load core: Plugin initialization failed: No themes found in installation.
```

<details>

```
/nix/store/88wqa2pardf08hlg0i4ayhka9rznzvm3-qtcreator-9.0.1/bin/qtcreator
qt.core.plugin.loader: In /nix/store/50jxrdi9hslmwph80drcvaw1q0ghnm96-qtwayland-5.15.7-bin/lib/qt-5.15.7/plugins/platforms/libqwayland-generic.so:
  Plugin uses incompatible Qt library (5.15.0) [release]
qt.core.plugin.loader: In /nix/store/50jxrdi9hslmwph80drcvaw1q0ghnm96-qtwayland-5.15.7-bin/lib/qt-5.15.7/plugins/platforms/libqwayland-xcomposite-egl.so:
  Plugin uses incompatible Qt library (5.15.0) [release]
...
Warning: No themes found in installation:  "/share/qtcreator/themes"
SOFT ASSERT [09:25:44.102]: "d" in /build/qt-creator-opensource-src-9.0.1/src/plugins/qmlprojectmanager/qmlprojectplugin.cpp:103
SOFT ASSERT [09:25:44.106]: "d" in /build/qt-creator-opensource-src-9.0.1/src/plugins/vcsbase/vcsplugin.cpp:53
SOFT ASSERT [09:25:44.108]: "dd" in /build/qt-creator-opensource-src-9.0.1/src/plugins/projectexplorer/projectexplorer.cpp:791
Failed to load core: Plugin initialization failed: No themes found in installation.
```

unset QT_PLUGIN_PATH to reduce noise

```
QT_PLUGIN_PATH= /nix/store/88wqa2pardf08hlg0i4ayhka9rznzvm3-qtcreator-9.0.1/bin/qtcreator
Warning: No themes found in installation:  "/share/qtcreator/themes"
SOFT ASSERT [09:12:29.319]: "d" in /build/qt-creator-opensource-src-9.0.1/src/plugins/qmlprojectmanager/qmlprojectplugin.cpp:103
SOFT ASSERT [09:12:29.321]: "d" in /build/qt-creator-opensource-src-9.0.1/src/plugins/vcsbase/vcsplugin.cpp:53
SOFT ASSERT [09:12:29.322]: "dd" in /build/qt-creator-opensource-src-9.0.1/src/plugins/projectexplorer/projectexplorer.cpp:791
Failed to load core: Plugin initialization failed: No themes found in installation.
```

> Warning: No themes found in installation:  "/share/qtcreator/themes"

should be

> /nix/store/88wqa2pardf08hlg0i4ayhka9rznzvm3-qtcreator-9.0.1/share/qtcreator/themes

[qtcreator/src/plugins/coreplugin/themechooser.cpp](https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/src/plugins/coreplugin/themechooser.cpp#n188)

```cc
#include "icore.h"
// ...
QList<ThemeEntry> ThemeEntry::availableThemes()
{
    QList<ThemeEntry> themes;

    static const FilePath installThemeDir = ICore::resourcePath("themes");
    static const FilePath userThemeDir = ICore::userResourcePath("themes");
    addThemesFromPath(installThemeDir.toString(), &themes);
    if (themes.isEmpty())
        qWarning() << "Warning: No themes found in installation: "
                   << installThemeDir.toUserOutput();
```

so `installThemeDir` is wrong

[qtcreator/src/plugins/coreplugin/icore.cpp](https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/src/plugins/coreplugin/icore.cpp#n422)

```cc
FilePath ICore::resourcePath(const QString &rel)
{
    return FilePath::fromString(
               QDir::cleanPath(QCoreApplication::applicationDirPath() + '/' + RELATIVE_DATA_PATH))
           / rel;
}
```

debug ...

```
ICore::resourcePath: QCoreApplication::applicationDirPath() =  "/nix/store/6mdgsacs3balbcdxxg6pjckf5y2wypjf-qtcreator-9.0.1/bin"
ICore::resourcePath: QDir::cleanPath(QCoreApplication::applicationDirPath()) =  "/nix/store/6mdgsacs3balbcdxxg6pjckf5y2wypjf-qtcreator-9.0.1/bin"
ICore::resourcePath: RELATIVE_DATA_PATH =  ../../../../share/qtcreator
ICore::resourcePath: rel =  "themes"
```

so `RELATIVE_DATA_PATH` is wrong

```
grep -r RELATIVE_DATA_PATH
qbs/modules/qtc/qtc.qbs:        'RELATIVE_DATA_PATH="' + FileInfo.relativePath('/' + ide_bin_path,
cmake/QtCreatorAPI.cmake:file(RELATIVE_PATH RELATIVE_DATA_PATH "/${IDE_BIN_PATH}" "/${IDE_DATA_PATH}")
cmake/QtCreatorAPI.cmake:  RELATIVE_DATA_PATH="${RELATIVE_DATA_PATH}"
qtcreator.pri:RELATIVE_DATA_PATH = $$relative_path($$IDE_DATA_PATH, $$IDE_BIN_PATH)
qtcreator.pri:DEFINES += $$shell_quote(RELATIVE_DATA_PATH=\"$$RELATIVE_DATA_PATH\")
```

[qtcreator/qbs/modules/qtc/qtc.qbs](https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/qbs/modules/qtc/qtc.qbs)

```
        'RELATIVE_DATA_PATH="' + FileInfo.relativePath('/' + ide_bin_path,
                                                       '/' + ide_data_path) + '"',
```

so `IDE_BIN_PATH` or `IDE_DATA_PATH` is wrong

which file is the source of truth?

- `qbs/modules/qtc/qtc.qbs`
- `cmake/QtCreatorAPI.cmake`
- `qtcreator.pri`

&rarr; `cmake/QtCreatorAPI.cmake`

debug ...

```
QtCreatorAPIInternal.cmake: CMAKE_INSTALL_BINDIR=/nix/store/r25hz4hxbvmpbx08mjrijwgzizas1f26-qtcreator-9.0.1/bin
QtCreatorAPIInternal.cmake: CMAKE_INSTALL_LIBDIR=/nix/store/r25hz4hxbvmpbx08mjrijwgzizas1f26-qtcreator-9.0.1/lib
QtCreatorAPIInternal.cmake: CMAKE_INSTALL_LIBEXECDIR=/nix/store/r25hz4hxbvmpbx08mjrijwgzizas1f26-qtcreator-9.0.1/libexec
QtCreatorAPIInternal.cmake: CMAKE_INSTALL_DATAROOTDIR=share
```

so `CMAKE_INSTALL_DATAROOTDIR` is wrong, should be absolute path

https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html

so this is a bug in nixpkgs cmake

nixpkgs/pkgs/development/tools/build-managers/cmake/setup-hook.sh

```sh
    # This ensures correct paths with multiple output derivations
    # It requires the project to use variables from GNUInstallDirs module
    # https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html
    cmakeFlags="-DCMAKE_INSTALL_BINDIR=${!outputBin}/bin $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_SBINDIR=${!outputBin}/sbin $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_INCLUDEDIR=${!outputInclude}/include $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_OLDINCLUDEDIR=${!outputInclude}/include $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_MANDIR=${!outputMan}/share/man $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_INFODIR=${!outputInfo}/share/info $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_DOCDIR=${!outputDoc}/share/doc/${shareDocName} $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_LIBDIR=${!outputLib}/lib $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_LIBEXECDIR=${!outputLib}/libexec $cmakeFlags"
    cmakeFlags="-DCMAKE_INSTALL_LOCALEDIR=${!outputLib}/share/locale $cmakeFlags"
```

&rarr; `CMAKE_INSTALL_DATAROOTDIR` is not set

quickfix ...

```
pkgs/games/warzone2100/default.nix:    # WZ_DATADIR is based on CMAKE_INSTALL_DATAROOTDIR, so we set that.
pkgs/games/warzone2100/default.nix:    "-DCMAKE_INSTALL_DATAROOTDIR=${placeholder "out"}/share"
```

```nix
{
  cmakeFlags = [
    # workaround for missing CMAKE_INSTALL_DATAROOTDIR
    # in pkgs/development/tools/build-managers/cmake/setup-hook.sh
    "-DCMAKE_INSTALL_DATAROOTDIR=${placeholder "out"}/share"
  ];
```

success! looks ugly but works. should use dark theme. should scale fonts

qtcreator-9 0 1-hello

create project &rarr; kit selection &rarr; no suitable kits found

qtcreator-9 0 1-create-project-no-suitable-kits-found

preferences &rarr; kits &rarr; desktop

qtcreator-9 0 1-preferences-kits-desktop-cmake-debugger-missing

TODO

- [x] find gcc compiler
- [x] find gdb debugger
- [ ] find clang compiler

<details>
<summary>
configure summary
</summary>

```
-- The following packages have been found:

 * Qt6CoreTools (required version >= 6.4.1)
 * Qt6Core (required version >= 6.2.0)
 * Qt6Concurrent
 * OpenGL
 * PkgConfig
 * XKB (required version >= 0.5.0), XKB API common to servers and clients., <http://xkbcommon.org>
 * Vulkan
 * Qt6GuiTools (required version >= 6.4.1)
 * Qt6DBusTools (required version >= 6.4.1)
 * Qt6Gui
 * Qt6Network
 * Qt6WidgetsTools (required version >= 6.4.1)
 * Qt6PrintSupport
 * Qt6QmlTools (required version >= 6.4.1)
 * Qt6Qml
 * Qt6Sql
 * Qt6Xml
 * Qt6Core5Compat
 * WrapVulkanHeaders
 * Qt6Quick
 * Qt6QuickWidgets
 * Qt6Test
 * yaml-cpp, using internal src/libs/3rdparty/yaml-cpp
 * Qt6ShaderToolsTools (required version >= 6.4.1)
 * Qt6ShaderTools
 * Googletest, Google Testing and Mocking Framework, <https://github.com/google/googletest>
 * Threads
 * Qt6
 * Qt5
 * Python3

-- The following packages have not been found:

 * Qt6QmlCompilerPlusPrivate
 * Qt6LinguistTools
 * Qt6DesignerComponentsPrivate
 * Qt6Designer
 * Qt6DesignerComponents
 * Qt6Help
 * Qt6SerialPort
 * Qt6Svg
 * Qt6Tools
 * Clang
 * litehtml
 * Qt6SvgWidgets
 * Qt6WebEngineWidgets
 * Qt6Quick3D
 * Qt6Quick3DAssetImport
 * Qt6Quick3DParticles
 * Qt6Quick3DAssetUtils
 * Qt6QuickTimeline
 * elfutils, a collection of utilities and libraries to read, create and modify ELF binary files, <https://sourceware.org/elfutils/>

-- The following features have been enabled:

 * Library 3rd_cplusplus
 * Library KSyntaxHighlighting
 * Library AdvancedDockingSystem
 * Library Aggregation
 * Library ExtensionSystem
 * Library Utils
 * Library LanguageUtils
 * Library CPlusPlus
 * Library Modeling
 * Library QmlJS
 * Library QmlDebug
 * Library QmlEditorWidgets
 * Library GLSL
 * Library LanguageServerProtocol
 * Library SqliteC, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0
 * Library Sqlite, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0
 * Library Tracing, with CONDITION TARGET Qt6::ShaderTools AND TARGET Qt5::Quick
 * Library ProParser
 * Plugin Core
 * Plugin TextEditor
 * Plugin HelloWorld
 * Plugin ImageViewer
 * Plugin Marketplace
 * Plugin UpdateInfo
 * Plugin Welcome
 * Plugin BinEditor
 * Plugin CodePaster
 * Plugin DiffEditor
 * Plugin EmacsKeys
 * Plugin Macros
 * Plugin ProjectExplorer
 * Plugin SilverSearcher
 * Plugin Bookmarks
 * Plugin CppEditor
 * Plugin ResourceEditor
 * Plugin Nim
 * Plugin Conan
 * Plugin ClassView
 * Plugin GLSLEditor
 * Plugin ModelEditor
 * Plugin QtSupport
 * Plugin Todo
 * Plugin VcsBase
 * Plugin Bazaar
 * Plugin Beautifier
 * Plugin ClearCase
 * Plugin CVS
 * Plugin Docker
 * Plugin FakeVim
 * Plugin GenericProjectManager
 * Plugin Git
 * Plugin Mercurial
 * Plugin MesonProjectManager
 * Plugin Perforce
 * Plugin QmakeProjectManager
 * Plugin QmlJSTools
 * Plugin QmlProjectManager, with CONDITION TARGET Qt5::QuickWidgets
 * Plugin ScxmlEditor
 * Plugin Subversion
 * Plugin CompilationDatabaseProjectManager
 * Plugin LanguageClient
 * Plugin CMakeProjectManager
 * Plugin Debugger
 * Plugin Coco
 * Plugin GitLab
 * Plugin Android
 * Plugin AutoTest
 * Plugin AutotoolsProjectManager
 * Plugin BareMetal
 * Plugin ClangCodeModel
 * Library yaml-cpp
 * Plugin ClangTools, with CONDITION TARGET yaml-cpp
 * Plugin Cppcheck
 * Plugin IncrediBuild
 * Plugin Ios
 * Plugin Python
 * Plugin QmlJSEditor
 * Plugin QmlPreview, with CONDITION TARGET QmlProjectManager
 * Plugin QmlProfiler, with CONDITION TARGET Tracing AND TARGET Qt6::ShaderTools
 * Plugin RemoteLinux
 * Plugin Valgrind
 * Plugin PerfProfiler, with CONDITION TARGET Tracing
 * Plugin QbsProjectManager
 * Plugin CtfVisualizer, with CONDITION TARGET Tracing
 * Plugin Squish
 * Plugin Boot2Qt
 * Plugin Qnx
 * Plugin WebAssembly
 * Plugin McuSupport
 * Library qml2puppet_static
 * Library sdktoolLib

-- The following features have been disabled:

 * Build documentation
 * Build online documentation
 * Build tests
 * Build with sanitize, SANITIZE_FLAGS=''
 * Build with Crashpad
 * Library Nanotrace
 * Build Qbs
 * Plugin SerialTerminal, with CONDITION TARGET Qt5::SerialPort
 * Plugin Help, with CONDITION TARGET Qt5::Help
 * Plugin ClangFormat, with CONDITION TARGET clangFormat AND LLVM_PACKAGE_VERSION VERSION_GREATER_EQUAL 10.0.0 AND ( QTC_CLANG_BUILDMODE_MATC>
 * Plugin Designer, with CONDITION TARGET Qt5::DesignerComponents AND TARGET Qt5::Designer
 * ProjectStorage
 * Build with QmlDom
 * Plugin QmlDesigner, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0 AND TARGET Qt5::QuickWidgets AND TARGET Qt5::Svg
 * Plugin assetexporterplugin, with CONDITION TARGET QmlDesigner
 * Plugin componentsplugin, with CONDITION TARGET QmlDesigner
 * Plugin qmlpreviewplugin, with CONDITION TARGET QmlDesigner
 * Plugin qtquickplugin, with CONDITION TARGET QmlDesigner
 * Plugin StudioPlugin, with CONDITION TARGET QmlDesigner
 * Plugin StudioWelcome, with CONDITION TARGET Qt5::QuickWidgets AND TARGET QmlDesigner
 * Qt Quick 3D support, with CONDITION TARGET Qt5::Quick3D
 * Qt Quick 3D asset import, with CONDITION TARGET Qt5::Quick3DAssetImport
 * Qt Quick 3D particles, with CONDITION TARGET Qt5::Quick3DParticles
 * Qt Quick 3D asset utils, with CONDITION TARGET Qt5::Quick3DAssetUtils
 * multilanguage-support in qml2puppet, with CONDITION TARGET QtCreator::multilanguage-support
 * Include developer documentation
```

</details>

</details>

**Comment by bjorn** [comment]
I recently started using NixOS and I'd love to use an up to date Qt Creator. Is there any way in which I can help finish this PR?

**Comment by milahu** [comment]
using qtcreator 9 is as simple as

```sh
cd ~/src
git clone --depth 1 https://github.com/NixOS/nixpkgs
cd nixpkgs
nix-shell -p nixpkgs-review
nixpkgs-review pr 189504
qtcreator
```

getting it merged into nixpkgs is harder. if that is your goal, you could ...

test qtcreator (basic functions, plugins, ...) and post your results here

ping some nixpkgs maintainers or post in https://discourse.nixos.org/t/prs-ready-for-review/3032

**Comment by bjorn** [comment]
Hmm, I'm getting the following error trying to use `nixpkgs-review`:
```
[nix-shell:~/playground/nixpkgs]$ nixpkgs-review pr 189504
$ git worktree prune
Traceback (most recent call last):
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/bin/.nixpkgs-review-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/__init__.py", line 10, in main
    cli.main(command, args)
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/cli/__init__.py", line 301, in main
    return cast(str, args.func(args))
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/cli/pr.py", line 68, in pr_command
    contexts.append((pr, builddir.path, review.build_pr(pr)))
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/review.py", line 199, in build_pr
    pr = self.github_client.pull_request(pr_number)
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/github.py", line 70, in pull_request
    return self.get(f"repos/NixOS/nixpkgs/pulls/{number}")
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/github.py", line 33, in get
    return self._request(path, "GET")
  File "/nix/store/28x296plr3khwyi163wgbyw77kw50v26-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/github.py", line 29, in _request
    resp = urllib.request.urlopen(req)
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/nix/store/sp5x6s8n36gjlwck74xhj1i61p66vcpa-python3-3.10.9/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unauthorized
```

Btw, it seems the update from 8.0.1 to 9.0.1 is still only on your beta branch and not on this PR.

**Comment by milahu** [comment]
> the update from 8.0.1 to 9.0.1 is still only on your beta branch and not on this PR.

fixed (but i cant change the branch name, so its still qtcreator-8.0.1)

> HTTP Error 401: Unauthorized

sounds like github api rate-limiting

you can also just

```
cd nixpkgs
git fetch https://github.com/milahu/nixpkgs qtcreator-8.0.1:qtcreator-8.0.1 --depth 26
git checkout qtcreator-8.0.1
nix-build . -A qtcreator
```

**Comment by bjorn** [comment]
Thanks @milahu, that worked for me!

Unfortunately I ran into other issues with setting up my development environment (cmake not finding xkb when depending on QtGui private, which was beyond me to solve), so for now I've switched back to a more traditional distribution.

**Comment by milahu** [comment]
> I've switched back to a more traditional distribution.

yepp, simpler for development

> cmake not finding xkb when depending on QtGui private

just add

```nix
{
  buildInputs = [
    libxkbcommon
  ];
```

but in theory it should work with

```nix
{
  buildInputs = [
    qt6.qtbase
  ];
```

because

https://github.com/NixOS/nixpkgs/blob/7a033d6395a500f53c03b808855d33a9aafc894b/pkgs/development/libraries/qt-6/modules/qtbase.nix#L155

```nix
{
  propagatedBuildInputs = [
    # ...
  ] ++ lib.optionals systemdSupport [
    systemd
  ] ++ lib.optionals stdenv.isLinux [
    # ...
    libxkbcommon
```

**Comment by bjorn** [comment]
@milahu Unfortunately, I have no idea why I was getting that error, but I definitely tried to get `libxkbcommon` in there manually, so it must have been something else. Also, I tried using `nix develop` on Fedora now to enter the development environment and am not running into that problem now. :-S

But I was somewhat reluctant to even mention it because I was afraid you'd try to help me with it. ;-) Unfortunately, even when this works now, I'm just running into the next problem, at linking time:
```
[ 98%] Building CXX object CMakeFiles/Notes.dir/Notes_autogen/UVLADIE3JM/qrc_styles.cpp.o
[100%] Linking CXX executable notes
/nix/store/r2b9k28c6aghczpqfvh71y9xavm7rr68-binutils-2.39/bin/ld: warning: libdl.so.2, needed by /nix/store/72wr6baiks8b500xz0dw6485dgxmbmvs-qtbase-6.4.2/lib/libQt6Network.so.6.4.2, not found (try using -rpath or -rpath-link)
/nix/store/r2b9k28c6aghczpqfvh71y9xavm7rr68-binutils-2.39/bin/ld: warning: libgssapi_krb5.so.2, needed by /nix/store/72wr6baiks8b500xz0dw6485dgxmbmvs-qtbase-6.4.2/lib/libQt6Network.so.6.4.2, not found (try using -rpath or -rpath-link)
/nix/store/r2b9k28c6aghczpqfvh71y9xavm7rr68-binutils-2.39/bin/ld: warning: libbrotlidec.so.1, needed by /nix/store/72wr6baiks8b500xz0dw6485dgxmbmvs-qtbase-6.4.2/lib/libQt6Network.so.6.4.2, not found (try using -rpath or -rpath-link)
... (followed by hundreds "undefined reference" errors as expected)
```
And I do not know what I'd need to do to "try using -rpath or -rpath-link".

**Comment by milahu** [comment]
> I tried using `nix develop` on Fedora

that is out of my scope ... i would guess that your compiler is not using the $NIX_* envs
maybe ask in matrix chat

**Comment by milahu** [comment]
qtcreator 9.0.1 fails to detect qtdeclarative

```sh
nix-shell -p cmake gnumake gcc gdb qt6.{qtbase,qttools,qtdeclarative}{,.dev} 
qml --version 2>/dev/null 
# Qml Runtime 6.4.2
./result/bin/qtcreator
```

edit, preferences, kits, qt versions, qt 6

> No QML utility installed

```sh
strace -f ./result/bin/qtcreator 2>qtcreator-trace.txt
```

> [pid 99461] access("/nix/store/6kj9nzxmvph2w30vwmmnsjpg9p8p0ifc-qtcreator-9.0.1/share/qtcreator/externaltools/qml.xml", R_OK) = 0

```xml
<externaltool id="qml">
    <description>Runs the current QML file with QML utility.</description>
    <displayname>QML utility</displayname>
    <category>Qt Quick</category>
    <order>1</order>
    <executable>
        <path>%{CurrentDocument:Project:QT_INSTALL_BINS}/qml</path>
        <path>qml</path>
        <arguments>%{CurrentDocument:FilePath}</arguments>
        <workingdirectory>%{CurrentDocument:Path}</workingdirectory>
        <environment>QT_LOGGING_TO_CONSOLE%3D1</environment>
    </executable>
</externaltool>
```

fails because `QT_INSTALL_BINS=${qt6.qtbase.dev}/bin`
and qml is installed in `${qt6.qtdeclarative.dev}/bin`

**Comment by domenkozar** [comment]
user was banned https://github.com/NixOS/moderation/blob/main/moderation-log.md

**Pull request closed without merging**